src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -190,6 +213,24 @@ balanceTxWithConstraints unbalancedTx constraintsBuilder = do
       =<< maybe (liftQueryM QueryM.getChangeAddress) (pure <<< Just)
       =<< asksConstraints Constraints._changeAddress
 
+  getWalletUtxosAt :: Array Address -> QueryM (Either BalanceTxError UtxoMap)
+  getWalletUtxosAt inWalletAddresses = do
+    filteredWalletUtxos <-
+      maybe (Map.empty)
+        ( Map.filter
+            (\txout -> elem (_.address (unwrap txout)) inWalletAddresses)
+        ) <$> getWalletUtxos
+    utxosAtWalletAddresses <- parallelUtxosAt inWalletAddresses
+    pure $ case Map.isSubmap filteredWalletUtxos <$> utxosAtWalletAddresses of
```

Comment:
I'm running into an issue with the logic here.
Specifically, now that `waitForInputs` is no longer used, when running AlwaysSucceeds with a newly initialized wallet (or wallet with 1 spendable UTxO) there is a scenario, after the initial payment to the alwaysSucceedScript, in which Eternl hasn't yet updated it's UTxO list and therefore `filteredWalletUtxos` is no longer a subset of `utxosAtWalletAddress` and so the example fails.

_I'm currently investigating further, will add more details and logs as I go forward_

Reply:
Yes, this is the bug I have been tracing all weekend, I came to the conclusion that using `getWalletUtxos` in the balancer once before beginning, has the effect of allowing the first test in the e2e suit to pass and I was suspecting it because it forced the wallet to update it' utxos. This problem is affecting `develop` right now.

Sadly it only works for the fist contract that is run in the suite, the rest of them fail after that, so I need to find proper way to force the update.

Reply:
I would try to add a waiting loop in the balancer and see if it helps

Reply:
It didn't help, I did something like `run this function once every second for 20 seconds and keep the last returned result`. It seems that the problem may be that the addresses are outdated. For some reason, I keep seeing that every transaction generates a new address for the wallet and all the assets that aren't locked in the contract are moved to the new address. While constructing the transaction the addresses are outdated and so the new address with all the assets isn't part of the return of `asksConstraints Constraints._srcAddresses` in the balancer, effectively hiding the utxos in this algorithm. I would test it tomorrow.

Reply:
With regards to the new addresses, from what I've seen, `getChangeAddress` generates a new address on every call. But the address is not returned from `getUsedAddresses` until the wallet can see the outputs of the Tx.

Any wallet that is multi-address will probably do this, I know Yoro certainly does.

Reply:
tanks! I was wondering if I misunderstood how addresses work.